runtime/protocol/aws-protocol-core/common/src/aws/smithy/kotlin/runtime/awsprotocol/ClockSkewInterceptor.kt

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.awsprotocol
+
+import aws.smithy.kotlin.runtime.ErrorMetadata
+import aws.smithy.kotlin.runtime.SdkBaseException
+import aws.smithy.kotlin.runtime.ServiceErrorMetadata
+import aws.smithy.kotlin.runtime.client.ProtocolRequestInterceptorContext
+import aws.smithy.kotlin.runtime.client.ResponseInterceptorContext
+import aws.smithy.kotlin.runtime.http.interceptors.HttpInterceptor
+import aws.smithy.kotlin.runtime.http.operation.HttpOperationContext
+import aws.smithy.kotlin.runtime.http.request.HttpRequest
+import aws.smithy.kotlin.runtime.http.response.HttpResponse
+import aws.smithy.kotlin.runtime.http.response.header
+import aws.smithy.kotlin.runtime.telemetry.logging.logger
+import aws.smithy.kotlin.runtime.time.Instant
+import aws.smithy.kotlin.runtime.time.until
+import aws.smithy.kotlin.runtime.util.get
+import kotlinx.atomicfu.*
+import kotlin.coroutines.coroutineContext
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.minutes
+
+/**
+ * An interceptor used to detect clock skew (difference between client and server clocks) and apply a correction.
+ */
+public class ClockSkewInterceptor : HttpInterceptor {
+    public companion object {
+        /**
+         * How much must the clock be skewed before attempting correction
+         */
+        public val CLOCK_SKEW_THRESHOLD: Duration = 4.minutes
+
+        /**
+         * Determine whether the client's clock is skewed relative to the server.
+         * @return true if the service's response represents a definite clock skew error
+         * OR a *possible* clock skew error AND the skew exists. false otherwise.
+         * @param responseCodeDescription the server's response code description
+         * @param serverTime the server's time
+         */
+        internal fun Instant.isSkewed(serverTime: Instant, responseCodeDescription: String?): Boolean =
```

Comment:
**nit**: `responseCodeDescription` -> `errorCode`

Also `responseCodeDescription` should probably be non-nullable since this always returns false otherwise which means it's really a required parameter for determining if a clock is skewed.